### ISSUE TITLE
Moderation

### ISSUE THREAD

**Issue by cycomachead** [bromagosa/snapCloud#110]
Moderation applies to all main objects, i.e. projects, users, and comments. 

We should break this up into multiple issues, at some point, but here's a high level idea:

* users can report an item. An item is a specific revision of a project or comment.
* all reports go into a queue for review.
* after 3 reports the item is automatically hidden.
* moderators can review reports and approve or deny them.
    * denying makes the item available again
    * approving keeps the item hidden, but removes it from the queue. Subsequent updates are still hidden.
   * there should be notes/a selector for why a request was approved or denied.
* people can request a review of their moderated content after it has been updated
   * it should ask them to explain what has changed
* moderation features need to track the person requesting and reviewing requests

This requires:
* a `reports` table
    * basically a list of flags that link objects and moderations
* a `moderations` table?
   * to contain info about the moderation and which version something links to
* eventually an `is_moderator` user option, that's not an admin
   * moderators should be able to see reported unpublished projects, but no private data
* a bunch of front-end UI
* `/report` endpoint for projects and so on
* a set of `moderation/X/` endpoints

**Comment by bromagosa** [comment]
I agree we'll eventually need moderation, but I don't think we need to have this done before the Public Release. We can (and I think we should) definitely release before even enabling comments. If we keep adding fuel to the fire we'll never go public :)

**Comment by cycomachead** [comment]
Perhaps we don’t need full moderation, right away, but I do think we need at least the ability for users to flag projects. We need some insight into preventing bad behavior, and we need to be able to tell teachers or IT folks that there’s an easy way to deal with bad content.

**Comment by bromagosa** [comment]
Flagging projects is a useful one, yep.

**Comment by bromagosa** [comment]
Since flagging is working and we now have user roles that can unpublish projects, I'm closing this meta issue. Let's open new issues for particular missing moderation features.

**Comment by cycomachead** [comment]
Sounds good!

-- 
Michael Ball
From my iPhone
michaelball.co
